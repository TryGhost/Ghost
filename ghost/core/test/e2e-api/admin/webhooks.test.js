const {agentProvider, fixtureManager, matchers} = require('../../utils/e2e-framework');
const {anyContentVersion, anyEtag, anyErrorId, anyObjectId, anyISODate, stringMatching, anyContentLength} = matchers;

const webhookMatcher = {
    id: anyObjectId,
    api_version: stringMatching(/v\d+\.\d+/),
    integration_id: anyObjectId,
    created_at: anyISODate,
    updated_at: anyISODate
};

describe('Webhooks API', function () {
    let agent;
    let createdWebhookId;
    let webhookData;

    before(async function () {
        agent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init('integrations');
        await agent.loginAsOwner();

        // create a webhook linked to a real integration
        webhookData = {
            event: 'test.create',
            target_url: 'http://example.com/webhooks/test/extra/1',
            name: 'test',
            secret: 'thisissecret',
            integration_id: fixtureManager.get('integrations', 0).id
        };
    });

    it('Can create a webhook', async function () {
        const {body} = await agent.post('/webhooks/')
            .body({webhooks: [webhookData]})
            .expectStatus(201)
            .matchHeaderSnapshot({
                // Note: No location header as there is no read method for webhooks
                etag: anyEtag,
                'content-version': anyContentVersion,
                'content-length': anyContentLength

            })
            .matchBodySnapshot({
                webhooks: [webhookMatcher]
            });

        // Store an id for use in future tests. Not the best pattern but does keep the tests readable.
        createdWebhookId = body.webhooks[0].id;
    });

    it('Fails nicely when adding a duplicate webhook', async function () {
        await agent.post('/webhooks/')
            .body({webhooks: [webhookData]})
            .expectStatus(422)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                errors: [{
                    id: anyErrorId
                }]
            });
    });

    it('Fails nicely when creating an orphaned webhook', async function () {
        await agent
            .post('/webhooks/')
            .body({webhooks: [{
                event: 'test.create',
                target_url: 'http://example.com/webhooks/test/extra/10',
                name: 'test',
                secret: 'thisissecret',
                integration_id: `fake-integration`
            }]})
            .expectStatus(422)
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            })
            .matchBodySnapshot({
                errors: [{
                    id: anyErrorId
                }]
            });
    });

    it('Can edit a webhook', async function () {
        await agent.put(`/webhooks/${createdWebhookId}/`)
            .body({
                webhooks: [{
                    name: 'Edit Test',
                    event: 'member.added',
                    target_url: 'https://example.com/new-member',
                    integration_id: 'ignore_me'
                }]
            })
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag,
                'content-version': anyContentVersion,
                'content-length': anyContentLength
            })
            .matchBodySnapshot({
                webhooks: [webhookMatcher]
            });
    });

    it('Cannot edit a non-existent webhook', async function () {
        await agent.put('/webhooks/abcd1234abcd1234abcd1234/')
            .body({
                webhooks: [{
                    name: 'Edit Test',
                    event: 'member.added',
                    target_url: 'https://example.com/new-member',
                    integration_id: 'ignore_me'
                }]
            })
            .expectStatus(404)
            .matchBodySnapshot({
                errors: [{
                    id: anyErrorId
                }]
            })
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            });
    });

    it('Can delete a webhook', async function () {
        await agent
            .delete(`/webhooks/${createdWebhookId}/`)
            .expectStatus(204)
            .expectEmptyBody()
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            });
    });

    it('Cannot delete a non-existent webhook', async function () {
        await agent
            .delete('/webhooks/abcd1234abcd1234abcd1234/')
            .expectStatus(404)
            .matchBodySnapshot({
                errors: [{
                    id: anyErrorId
                }]
            })
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                etag: anyEtag
            });
    });
});
