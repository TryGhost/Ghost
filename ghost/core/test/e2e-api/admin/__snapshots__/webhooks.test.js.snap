// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhooks API Can create a webhook 1: [body] 1`] = `
Object {
  "webhooks": Array [
    Object {
      "api_version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}/,
      "event": "test.create",
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "integration_id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "last_triggered_at": null,
      "last_triggered_error": null,
      "last_triggered_status": null,
      "name": "test",
      "secret": "thisissecret",
      "target_url": "http://example.com/webhooks/test/extra/1",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}/,
    },
  ],
}
`;

exports[`Webhooks API Can create a webhook 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": StringMatching /\\\\d\\+/,
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Webhooks API Can delete a webhook 1: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin",
  "x-powered-by": "Express",
}
`;

exports[`Webhooks API Can edit a webhook 1: [body] 1`] = `
Object {
  "webhooks": Array [
    Object {
      "api_version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}/,
      "event": "member.added",
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "integration_id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "last_triggered_at": null,
      "last_triggered_error": null,
      "last_triggered_status": null,
      "name": "Edit Test",
      "secret": "thisissecret",
      "target_url": "https://example.com/new-member",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}/,
    },
  ],
}
`;

exports[`Webhooks API Can edit a webhook 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": StringMatching /\\\\d\\+/,
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Webhooks API Cannot delete a non-existent webhook 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Resource could not be found.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Resource not found error, cannot delete webhook.",
      "property": null,
      "type": "NotFoundError",
    },
  ],
}
`;

exports[`Webhooks API Cannot delete a non-existent webhook 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "260",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Webhooks API Cannot delete a nonexistent webhook 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Resource could not be found.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Resource not found error, cannot delete webhook.",
      "property": null,
      "type": "NotFoundError",
    },
  ],
}
`;

exports[`Webhooks API Cannot delete a nonexistent webhook 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "260",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Webhooks API Cannot edit a non-existent webhook 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Resource could not be found.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Resource not found error, cannot edit webhook.",
      "property": null,
      "type": "NotFoundError",
    },
  ],
}
`;

exports[`Webhooks API Cannot edit a non-existent webhook 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "258",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Webhooks API Cannot edit a nonexistent webhook 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Resource could not be found.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Resource not found error, cannot edit webhook.",
      "property": null,
      "type": "NotFoundError",
    },
  ],
}
`;

exports[`Webhooks API Cannot edit a nonexistent webhook 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "258",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Webhooks API Fails nicely when adding a duplicate webhook 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Target URL has already been used for this event.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save webhook.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Webhooks API Fails nicely when adding a duplicate webhook 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "272",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Webhooks API Fails nicely when creating an orphaned webhook 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Validation failed for 'integration_id'. 'integration_id' value does not match any existing integration.",
      "details": null,
      "ghostErrorCode": null,
      "help": "Provide the 'integration_id' of an existing integration.",
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save webhook.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Webhooks API Fails nicely when creating an orphaned webhook 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "381",
  "content-type": "application/json; charset=utf-8",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Accept-Version, Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
